src/main/scala/firrtl/passes/ExpandWhens.scala

```diff
@@ -140,10 +140,15 @@ object ExpandWhens extends Pass {
                   conseqNetlist.getOrElse(lvalue, altNetlist(lvalue))
               }
 
-              val memoNode = DefNode(s.info, namespace.newTemp, res)
-              val memoExpr = WRef(memoNode.name, res.tpe, NodeKind(), MALE)
-              memos += memoNode
-              netlist(lvalue) = memoExpr
+              res match {
+                case _: WInvalid|_: WRef|_: WSubField|_: WSubIndex|_: Literal =>
+                  netlist(lvalue) = res
+                case _ =>
+                  val memoNode = DefNode(s.info, namespace.newTemp, res)
+                  val memoExpr = WRef(memoNode.name, res.tpe, NodeKind(), MALE)
+                  memos += memoNode
+                  netlist(lvalue) = memoExpr
+              }
```

Comment:
A comment at the top of this match statement saying something like "Only memoize non-referential expressions" would be nice.

I also think it might be better to match on the expressions we *are* memoizing (ie. Muxes and Primops) rather than the ones we are not.